### ISSUE TITLE
preproccesing code

### ISSUE THREAD

**Issue by Phani123-tech** [Phani123-tech/cc#1]
#!/usr/bin/env python3
"""
ADAPT-SDN-Guard: Enhanced Preprocessing Script
Fixes bugs and adds feature selection + class balancing
"""

import pandas as pd
import numpy as np
from sklearn.preprocessing import StandardScaler, RobustScaler
from sklearn.model_selection import train_test_split
from sklearn.feature_selection import VarianceThreshold, SelectKBest, f_classif
import pickle
import os
import glob
import sys
from pathlib import Path

print("="*80)
print("ADAPT-SDN-GUARD: ENHANCED DATASET PREPROCESSING")
print("="*80)

# Create directories
os.makedirs('datasets/processed', exist_ok=True)
os.makedirs('results/models', exist_ok=True)
os.makedirs('results/logs', exist_ok=True)

# ============================================================================
# CONFIGURATION
# ============================================================================

DATASET_PATHS = {
    'InSDN': 'datasets/raw/InSDN/*.csv',
    'CICIDS2017': 'datasets/raw/CICIDS2017/*.csv',
    'CICIDS2018': 'datasets/raw/CICIDS2018/*.csv'
}

# Feature selection parameters
USE_FEATURE_SELECTION = True
MAX_FEATURES = 50  # Reduce from all features to top 50
VARIANCE_THRESHOLD = 0.01  # Remove low-variance features

# Class balancing
USE_CLASS_BALANCING = True
BALANCE_METHOD = 'undersample'  # 'undersample' or 'smote'

# ============================================================================
# HELPER FUNCTIONS
# ============================================================================

def load_dataset(pattern, dataset_name, max_rows=100000):
    """Load CSV files with error handling"""
    print(f"\n{'='*80}")
    print(f"Processing: {dataset_name}")
    print(f"{'='*80}")
    
    files = glob.glob(pattern)
    
    if not files:
        print(f"⚠ No files found at: {pattern}")
        return None
    
    print(f"Found {len(files)} file(s)")
    
    dfs = []
    for file in files[:5]:  # Limit to 5 files
        try:
            print(f"  Loading: {os.path.basename(file)[:40]}...", end=" ")
            
            # Try reading with different encodings
            try:
                df = pd.read_csv(file, encoding='utf-8', low_memory=False, nrows=max_rows)
            except:
                try:
                    df = pd.read_csv(file, encoding='latin-1', low_memory=False, nrows=max_rows)
                except:
                    df = pd.read_csv(file, encoding='iso-8859-1', low_memory=False, nrows=max_rows)
            
            # Clean column names
            df.columns = df.columns.str.strip()
            dfs.append(df)
            print(f"✓ ({len(df):,} rows)")
        except Exception as e:
            print(f"✗ Error: {str(e)[:40]}")
    
    if not dfs:
        return None
    
    combined = pd.concat(dfs, ignore_index=True)
    print(f"\n✓ Total loaded: {len(combined):,} rows")
    print(f"  Columns: {len(combined.columns)}")
    return combined

def extract_features(df, dataset_name):
    """Extract features and labels from dataframe - FIXED"""
    print(f"\nExtracting features from {dataset_name}...")
    
    # Find label column
    label_cols = ['Label', 'label', ' Label', 'Attack', 'attack', 'Class', 'class']
    label_col = None
    for col in label_cols:
        if col in df.columns:
            label_col = col
            break
    
    if label_col is None:
        print(f"  ⚠ Using last column as label")
        label_col = df.columns[-1]
    
    print(f"  Label column: '{label_col}'")
    
    # Extract labels - handle different formats
    y = []
    for val in df[label_col]:
        val_str = str(val).upper().strip()
        is_attack = any(attack in val_str for attack in ['DDOS', 'DOS', 'ATTACK', 'MALICIOUS'])
        y.append(1 if is_attack else 0)
    
    y = np.array(y)
    print(f"  Benign: {(y==0).sum():,} | Attack: {(y==1).sum():,}")
    
    # Select numeric features CORRECTLY
    numeric_cols = []
    for col in df.columns:
        if col != label_col:
            # FIXED: Actually check if column is numeric
            if pd.api.types.is_numeric_dtype(df[col]):
                numeric_cols.append(col)
            else:
                # Try converting non-numeric to numeric
                try:
                    converted = pd.to_numeric(df[col], errors='coerce')
                    if converted.notna().sum() > len(df) * 0.5:  # At least 50% valid
                        numeric_cols.append(col)
                except:
                    pass
    
    print(f"  Found {len(numeric_cols)} numeric features")
    
    # Extract features using only numeric columns
    X_df = df[numeric_cols].copy()
    
    # Clean data
    X_df = X_df.replace([np.inf, -np.inf], np.nan)
    
    # Convert to numeric
    for col in X_df.columns:
        X_df[col] = pd.to_numeric(X_df[col], errors='coerce')
    
    # Fill NaN with median (better than 0 for skewed data)
    X_df = X_df.fillna(X_df.median())
    
    # Convert to numpy array
    X = X_df.values
    
    print(f"  Features shape: {X.shape}")
    
    return X, y, list(X_df.columns)

def remove_low_variance_features(X, y, feature_names, threshold=0.01):
    """Remove features with low variance"""
    print(f"\nRemoving low-variance features (threshold={threshold})...")
    
    selector = VarianceThreshold(threshold=threshold)
    X_selected = selector.fit_transform(X)
    
    mask = selector.get_support()
    selected_features = [name for name, selected in zip(feature_names, mask) if selected]
    
    removed = len(feature_names) - len(selected_features)
    print(f"  Removed {removed} features")
    print(f"  Remaining: {len(selected_features)} features")
    
    return X_selected, selected_features

def select_best_features(X, y, feature_names, k=50):
    """Select k best features using ANOVA F-test"""
    print(f"\nSelecting top {k} features using ANOVA F-test...")
    
    if X.shape[1] <= k:
        print(f"  Already have {X.shape[1]} features (<= {k}), skipping")
        return X, feature_names
    
    selector = SelectKBest(f_classif, k=k)
    X_selected = selector.fit_transform(X, y)
    
    mask = selector.get_support()
    selected_features = [name for name, selected in zip(feature_names, mask) if selected]
    
    # Get feature scores
    scores = selector.scores_
    feature_scores = sorted(zip(feature_names, scores, mask), 
                          key=lambda x: x[1], 
                          reverse=True)
    
    print(f"  Top 10 features by score:")
    for i, (name, score, selected) in enumerate(feature_scores[:10], 1):
        marker = "✓" if selected else "✗"
        print(f"    {i}. {marker} {name[:40]}: {score:.2f}")
    
    return X_selected, selected_features

def balance_classes(X, y, method='undersample'):
    """Balance classes to improve model performance"""
    print(f"\nBalancing classes using {method}...")
    
    unique, counts = np.unique(y, return_counts=True)
    print(f"  Before: Benign={counts[0]:,}, Attack={counts[1]:,}")
    
    if method == 'undersample':
        # Random undersampling of majority class
        min_class_size = counts.min()
        
        indices_0 = np.where(y == 0)[0]
        indices_1 = np.where(y == 1)[0]
        
        if len(indices_0) > len(indices_1):
            # Undersample class 0
            selected_0 = np.random.choice(indices_0, len(indices_1), replace=False)
            selected_indices = np.concatenate([selected_0, indices_1])
        else:
            # Undersample class 1
            selected_1 = np.random.choice(indices_1, len(indices_0), replace=False)
            selected_indices = np.concatenate([indices_0, selected_1])
        
        X_balanced = X[selected_indices]
        y_balanced = y[selected_indices]
        
    elif method == 'smote':
        try:
            from imblearn.over_sampling import SMOTE
            smote = SMOTE(random_state=42)
            X_balanced, y_balanced = smote.fit_resample(X, y)
        except ImportError:
            print("  ⚠ SMOTE not available, using undersample instead")
            print("  Install: pip install imbalanced-learn")
            return balance_classes(X, y, method='undersample')
    else:
        return X, y
    
    unique, counts = np.unique(y_balanced, return_counts=True)
    print(f"  After: Benign={counts[0]:,}, Attack={counts[1]:,}")
    
    return X_balanced, y_balanced

# ============================================================================
# MAIN PROCESSING
# ============================================================================

all_X = []
all_y = []
all_feature_names = []
dataset_info = {}

for dataset_name, pattern in DATASET_PATHS.items():
    df = load_dataset(pattern, dataset_name, max_rows=100000)
    
    if df is not None:
        X, y, feature_names = extract_features(df, dataset_name)
        
        all_X.append(X)
        all_y.append(y)
        if not all_feature_names:  # Store first dataset's feature names
            all_feature_names = feature_names
        
        dataset_info[dataset_name] = {
            'samples': len(X),
            'features': len(feature_names),
            'normal': (y==0).sum(),
            'attack': (y==1).sum()
        }
    else:
        print(f"⚠ Skipping {dataset_name} - no files found")

if not all_X:
    print("\n✗ ERROR: No datasets loaded!")
    sys.exit(1)

# ============================================================================
# COMBINE DATASETS
# ============================================================================

print(f"\n{'='*80}")
print("COMBINING DATASETS")
print(f"{'='*80}")

# Use all features from each dataset (they may differ)
max_features = max(X.shape[1] for X in all_X)
print(f"Max features across datasets: {max_features}")

# Align all datasets to have same number of features
aligned_X = []
for X in all_X:
    if X.shape[1] < max_features:
        # Pad with zeros
        padding = np.zeros((X.shape[0], max_features - X.shape[1]))
        X_padded = np.hstack([X, padding])
    else:
        X_padded = X[:, :max_features]
    aligned_X.append(X_padded)

X_combined = np.vstack(aligned_X)
y_combined = np.hstack(all_y)

print(f"\nCombined Dataset:")
print(f"  Total samples: {len(X_combined):,}")
print(f"  Features: {X_combined.shape[1]}")
print(f"  Benign: {(y_combined==0).sum():,} ({100*(y_combined==0).sum()/len(y_combined):.1f}%)")
print(f"  Attack: {(y_combined==1).sum():,} ({100*(y_combined==1).sum()/len(y_combined):.1f}%)")

# Generate feature names for combined dataset
combined_feature_names = [f"Feature_{i}" for i in range(X_combined.shape[1])]

# ============================================================================
# FEATURE SELECTION
# ============================================================================

if USE_FEATURE_SELECTION:
    print(f"\n{'='*80}")
    print("FEATURE SELECTION")
    print(f"{'='*80}")
    
    # Step 1: Remove low-variance features
    X_combined, combined_feature_names = remove_low_variance_features(
        X_combined, y_combined, combined_feature_names, 
        threshold=VARIANCE_THRESHOLD
    )
    
    # Step 2: Select best features
    X_combined, combined_feature_names = select_best_features(
        X_combined, y_combined, combined_feature_names,
        k=min(MAX_FEATURES, X_combined.shape[1])
    )

# ============================================================================
# CLASS BALANCING
# ============================================================================

if USE_CLASS_BALANCING:
    print(f"\n{'='*80}")
    print("CLASS BALANCING")
    print(f"{'='*80}")
    
    X_combined, y_combined = balance_classes(
        X_combined, y_combined, 
        method=BALANCE_METHOD
    )

# ============================================================================
# NORMALIZATION (Use RobustScaler for better outlier handling)
# ============================================================================

print(f"\n{'='*80}")
print("FEATURE NORMALIZATION")
print(f"{'='*80}")

# RobustScaler is better for data with outliers
scaler = RobustScaler()
X_scaled = scaler.fit_transform(X_combined)

print(f"✓ Features normalized using RobustScaler (better for outliers)")
print(f"  Shape: {X_scaled.shape}")

# ============================================================================
# TRAIN-TEST SPLIT
# ============================================================================

print(f"\n{'='*80}")
print("TRAIN-TEST SPLIT (70:30 Stratified)")
print(f"{'='*80}")

X_train, X_test, y_train, y_test = train_test_split(
    X_scaled, y_combined,
    test_size=0.3,
    random_state=42,
    stratify=y_combined
)

print(f"\nTraining Set:")
print(f"  Samples: {len(X_train):,}")
print(f"  Benign: {(y_train==0).sum():,} ({100*(y_train==0).sum()/len(y_train):.1f}%)")
print(f"  Attack: {(y_train==1).sum():,} ({100*(y_train==1).sum()/len(y_train):.1f}%)")

print(f"\nTest Set:")
print(f"  Samples: {len(X_test):,}")
print(f"  Benign: {(y_test==0).sum():,} ({100*(y_test==0).sum()/len(y_test):.1f}%)")
print(f"  Attack: {(y_test==1).sum():,} ({100*(y_test==1).sum()/len(y_test):.1f}%)")

# ============================================================================
# SAVE PROCESSED DATA
# ============================================================================

print(f"\n{'='*80}")
print("SAVING PROCESSED DATA")
print(f"{'='*80}")

# Save numpy arrays
np.save('datasets/processed/X_train_fusion.npy', X_train)
np.save('datasets/processed/X_test_fusion.npy', X_test)
np.save('datasets/processed/y_train_fusion.npy', y_train)
np.save('datasets/processed/y_test_fusion.npy', y_test)
np.save('datasets/processed/X_fusion.npy', X_scaled)
np.save('datasets/processed/y_fusion.npy', y_combined)

print(f"\n✓ NumPy arrays saved:")
print(f"  - X_train_fusion.npy {X_train.shape}")
print(f"  - X_test_fusion.npy {X_test.shape}")
print(f"  - y_train_fusion.npy {y_train.shape}")
print(f"  - y_test_fusion.npy {y_test.shape}")
print(f"  - X_fusion.npy {X_scaled.shape}")
print(f"  - y_fusion.npy {y_combined.shape}")

# Save scaler
with open('results/models/scaler_fusion.pkl', 'wb') as f:
    pickle.dump(scaler, f)
print(f"✓ Scaler saved: results/models/scaler_fusion.pkl")

# Save feature names
feature_names_file = 'datasets/processed/feature_names_fusion.txt'
with open(feature_names_file, 'w') as f:
    f.write('\n'.join([f"{i+1}. {name}" for i, name in enumerate(combined_feature_names)]))
print(f"✓ Feature names saved: {feature_names_file}")

# Save dataset info
with open('results/logs/dataset_info.txt', 'w') as f:
    f.write("DATASET INFORMATION\n" + "="*70 + "\n\n")
    for name, info in dataset_info.items():
        f.write(f"{name}:\n")
        f.write(f"  Samples: {info['samples']:,}\n")
        f.write(f"  Features: {info['features']}\n")
        f.write(f"  Benign: {info['normal']:,}\n")
        f.write(f"  Attack: {info['attack']:,}\n\n")
    
    f.write("\nFINAL PROCESSED DATA\n" + "="*70 + "\n")
    f.write(f"Features after selection: {len(combined_feature_names)}\n")
    f.write(f"Total samples: {len(X_combined):,}\n")
    f.write(f"Class balance: Benign={np.sum(y_combined==0):,}, Attack={np.sum(y_combined==1):,}\n")

print(f"✓ Dataset info saved: results/logs/dataset_info.txt")

# ============================================================================
# SUMMARY
# ============================================================================

print(f"\n{'='*80}")
print("PREPROCESSING COMPLETE!")
print(f"{'='*80}")

print(f"\nSummary:")
print(f"  Datasets processed: {len(dataset_info.keys())}")  # FIXED: was all_datasets
print(f"  Total samples: {len(X_combined):,}")
print(f"  Features: {X_combined.shape[1]}")
print(f"  Train: {len(X_train):,} | Test: {len(X_test):,}")
print(f"\nEnhancements Applied:")
print(f"  ✓ Feature selection: {USE_FEATURE_SELECTION}")
print(f"  ✓ Class balancing: {USE_CLASS_BALANCING} ({BALANCE_METHOD})")
print(f"  ✓ RobustScaler normalization")
print(f"\n✓ Ready for model training!")
print(f"{'='*80}\n")
